test/unit/org/apache/cassandra/auth/FunctionResourceTest.java

```diff
@@ -0,0 +1,106 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.auth;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.junit.Test;
+
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.marshal.TypeParser;
+
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.junit.Assert.assertEquals;
+
+public class FunctionResourceTest
+{
+    private static final String ks = "fr_ks";
+    private static final String func = "functions";
+    private static final String name = "concat";
+    private static final String varType = "org.apache.cassandra.db.marshal.UTF8Type";
+
+    @Test
+    public void testFunction() throws Exception
+    {
+        FunctionResource expected = FunctionResource.root();
+        FunctionResource actual = FunctionResource.fromName(func);
+        assertEquals(expected, actual);
+        assertEquals(expected.getName(), actual.getName());
+    }
+
+    @Test
+    public void testFunctionKeyspace() throws Exception
+    {
+        FunctionResource expected = FunctionResource.keyspace(ks);
+        FunctionResource actual = FunctionResource.fromName(String.format("%s/%s", func, ks));
+        assertEquals(expected, actual);
+        assertEquals(expected.getKeyspace(), actual.getKeyspace());
+    }
+
+    @Test
+    public void testFunctionWithSingleInputParameter() throws Exception
+    {
+        List<AbstractType<?>> argTypes = new ArrayList<>();
+        argTypes.add(TypeParser.parse(varType));
+        FunctionResource expected = FunctionResource.function(ks, name, argTypes);
+        FunctionResource actual = FunctionResource.fromName(String.format("%s/%s/%s[%s]", func, ks, name, varType));
+        assertEquals(expected, actual);
+        assertEquals(expected.getKeyspace(), actual.getKeyspace());
+    }
+
+    @Test
+    public void testFunctionWithMultipleInputParameter() throws Exception
+    {
+        List<AbstractType<?>> argTypes = new ArrayList<>();
+        argTypes.add(TypeParser.parse(varType));
+        argTypes.add(TypeParser.parse(varType));
+        FunctionResource expected = FunctionResource.function(ks, name, argTypes);
+        FunctionResource actual = FunctionResource.fromName(String.format("%s/%s/%s[%s^%s]", func, ks, name, varType, varType));
+        assertEquals(expected, actual);
+        assertEquals(expected.getKeyspace(), actual.getKeyspace());
+    }
+
+    @Test
+    public void testFunctionWithoutInputParameter() throws Exception
+    {
+        List<AbstractType<?>> argTypes = new ArrayList<>();
+        FunctionResource expected = FunctionResource.function(ks, name, argTypes);
+        FunctionResource actual = FunctionResource.fromName(String.format("%s/%s/%s[]", func, ks, name));
```

Comment:
I'm not sure what behavior is supposed to be, but I guess a user may want to input the function without empty square brackets:
```
%s/%s/%s
```
I think we need to add a corresponding test and either support such behavior or start raising a clear error.

Reply:
I guess with the current logic such input is considered to be valid.

Reply:
Added, right it works.